### PR TITLE
Fixes #10106 - Make tests run in parallel

### PR DESCRIPTION
**Pull request by dLobatog**
This should force our test runs to use all available cores in the machine and create a process/database/cache file store for each of them. Each process will only run a subset of all tests, and all of these will run in parallel. 

In order to run tests in parallel and test this PR you would need to change database.yaml to:

```
test:
  database: yourproject_test<%= ENV['TEST_ENV_NUMBER'] %>
```

and run:

`rake parallel:create` <- to create the databases
`rake parallel:migrate` <- equivalent to db:test:migrate

`rake parallel:test` will run the tests. In my machine (4 processes), total completion time went down from nearly 10 minutes to around 5 minutes. I would expect it to be more like 2.5 minutes so I will investigate how to make it faster.

### CODE REVIEW

**Comment by unorthodoxgeek** [comment]
it's never going to be 4x faster, unless run on actual four machines

File: config/environments/test.rb [hunk 1]
```diff
@@ -47,4 +47,11 @@
 
   #enables a few aliases - context, should, and should_eventually methods
   config.minitest_spec_rails.mini_shoulda = true
+
+  # High logger level results in less IOps & faster tests
+  config.log_level = :warn
+  I18n.enforce_available_locales = false
```
Comment (domcleal) [inline config/environments/test.rb]: We can't set this unilaterally as our RPMs don't have it, see http://projects.theforeman.org/issues/4274.  Perhaps check if it's available, or handle this separately under that ticket.

Comment (dLobatog) [inline config/environments/test.rb]: No problem, I'll remove it, it was just to remove 4x the warning clutter

**Comment by dLobatog** [comment]
Updated after removing the I18n enforce evailable locales setting

File: config/environments/test.rb [hunk 2]
```diff
@@ -47,4 +47,10 @@
 
   #enables a few aliases - context, should, and should_eventually methods
   config.minitest_spec_rails.mini_shoulda = true
+
+  # High logger level results in less IOps & faster tests
+  config.log_level = :warn
+
+  # Use separate cache stores for parallel_tests
+  config.cache_store = :file_store, Rails.root.join("tmp", "cache", "paralleltests#{ENV['TEST_ENV_NUMBER']}")
```
Comment (domcleal) [inline config/environments/test.rb:52]: Might this conflict with foreman_memcache?  Would the latter take precedence, and would we need to change its default config to follow TEST_ENV_NUMBER?

File: test/functional/api/v2/users_controller_test.rb [hunk 3]
```diff
@@ -3,7 +3,7 @@
 class Api::V2::UsersControllerTest < ActionController::TestCase
   valid_attrs = { :login => "johnsmith",
                   :mail => 'john@example.com',
-                  :auth_source_id => AuthSourceInternal.first.id,
+                  :auth_source_id => AuthSourceInternal.first,
```
Comment (domcleal) [inline test/functional/api/v2/users_controller_test.rb]: Why this change?

File: test/integration/architecture_test.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 require 'test_helper'
 
-class ArchitectureTest < ActionDispatch::IntegrationTest
+class ArchitectureIntegrationTest < ActionDispatch::IntegrationTest
```
Comment (domcleal) [inline test/integration/architecture_test.rb:3]: Why, or is this an unrelated tidyup?

**Comment by domcleal** [comment]
Is this reliable for you?  I don't _think_ (though haven't double checked) I have any local failures running tests, but I got a few on this branch:

<pre>
  1) Failure:
Api::V2::UsersControllerTest#test_0005_should update user [/home/dcleal/code/foreman/foreman/test/functional/api/v2/users_controller_test.rb:45]:
Expected response to be a <:success>, but was <422>
  2) Failure:
ReportsTest#test_0001_reports:daily sends mail [/home/dcleal/code/foreman/foreman/test/lib/tasks/reports_test.rb:22]:
Failed assertion, no message given.
  3) Failure:
ReportsTest#test_0002_reports:daily shows a recent report [/home/dcleal/code/foreman/foreman/test/lib/tasks/reports_test.rb:37]:
Failed assertion, no message given.
</pre>

File: test/integration/architecture_test.rb (hunk 4 above)
Comment (dLobatog) [inline test/integration/architecture_test.rb:3]: `parallel_tests` loads all classes at some point (before splitting the load I guess), so this name would clash with the unit test ArchitectureTest

File: config/environments/test.rb (hunk 2 above)
Comment (dLobatog) [inline config/environments/test.rb:52]: No idea, but this is only applied to test.rb so why would memcache be needed?

Comment (domcleal) [inline config/environments/test.rb:52]: I was trying to work out if you have the plugin installed, shouldn't we let it use that rather than overriding the cache store here?

**Comment by domcleal** [comment]
@eLobato please see my last comment?

**Comment by dLobatog** [comment]
@domcleal 100% reliable so far locally, 5 tries, 0 failures. I've rebased this PR against develop just in case, can you try that again?

File: config/environments/test.rb [hunk 5]
```diff
@@ -47,4 +47,10 @@
 
   #enables a few aliases - context, should, and should_eventually methods
   config.minitest_spec_rails.mini_shoulda = true
+
+  # High logger level results in less IOps & faster tests
+  config.log_level = :warn
```
Comment (domcleal) [inline config/environments/test.rb]: We also have the test log level set in logging.yaml now, need to work out which one should be in use and if we should change it.

**Comment by domcleal** [comment]
Seems better, thanks.  I had a failure on the HostMailerTest strangely, but I don't know why (expectation's backwards again):

<pre>
  1) Failure:
HostMailerTest#test_0006_error_state sends mail with correct headers [/home/dcleal/code/foreman/foreman/test/unit/host_mailer_test.rb:58]:
Expected ["Foreman-noreply@redhat.com"] to include nil.
</pre>

It's not a big enough concern for me to not merge it, but I wonder why it happens to me anyway!

There are some inline comments above to address too.

**Comment by dLobatog** [comment]
@domcleal I think you're hitting a race condition right there, where `Setting` has not been loaded yet in the process running the assertion. I've changed it to reload Setting::General if it doesn't find it.

foreman_memcache has the potential to break tests here, but I don't think that should be an issue - I could try harder and force to set `config.cache_store` to `file_store` after foreman_memcache is loaded, but I don't think the plugin is meant to be run for tests.

**Comment by domcleal** [comment]
> @domcleal I think you're hitting a race condition right there, where Setting has not been loaded yet in the process running the assertion. I've changed it to reload Setting::General if it doesn't find it.

Sorry, would you mind just explaining that again?  I don't follow.  I think settings are only loaded via fixtures during test runs, due to DB cleaner - so I'd have thought the value would be the one coming from test/fixtures/settings.yml in all cases.

The only thing that might confuse it is the cache - retrieving a setting value does use the Rails cache, so that was being shared somehow (and I understand it oughtn't be) then it might cause strange results.

Still, it's not a terribly useful assertion!

**Comment by dLobatog** [comment]
@domcleal I went through this again. What I think it's happening is:
- load_defaults is only ran by one process (out of the 4) via config/initializers/foreman.rb (or something else?)
- Tests are given to each process randomly, so that if the HostMailer test is given to a process that hasn't ran load_defaults, `Setting['email_reply_address']` will return nil. 

Adding a fixture does help as fixtures are loaded in all 4 processes. If the fixture value is set to `"Foreman-noreply@#{domain}"`, the default setting, it'd work too.  Do you prefer that approach?

**Comment by domcleal** [comment]
> Adding a fixture does help as fixtures are loaded in all 4 processes. If the fixture value is set to "Foreman-noreply@#{domain}", the default setting, it'd work too. Do you prefer that approach?

Yep, then I think that matches more closely to how we use setting defaults from fixtures elsewhere in tests.  I'm a little surprised that load_defaults still has any effect on tests though.

**Comment by dLobatog** [comment]
@domcleal  Cool, I updated the fixture, it seems green after that. Thanks!

**Comment by domcleal** [comment]
Thanks @eLobato, merged as 9aadd81.  I'd suggest adding this to the development resources section of the wiki.
